File: training-data/ocaml/selection_sort.ml
let rec selection_sort = function
  | [] -> []
  | lst ->
      let min_val = List.fold_left min (List.hd lst) lst in
      let rest = List.filter (fun x -> x <> min_val) lst in
      min_val :: selection_sort rest

let () =
  selection_sort [ 5; 3; 8; 1; 9; 2 ]
  |> List.map string_of_int
  |> String.concat " "
  |> print_endline

File: training-data/prolog/selection_sort.pl
select_min(X, [], X, []).
select_min(X, [H|T], Min, [X|R]) :-
    X > H, !,
    select_min(H, T, Min, R).
select_min(X, [H|T], Min, [H|R]) :-
    select_min(X, T, Min, R).

selection_sort([], []).
selection_sort([H|T], [Min|Sorted]) :-
    select_min(H, T, Min, Rest),
    selection_sort(Rest, Sorted).

:- selection_sort([5,3,8,1,9,2], Sorted), writeln(Sorted).
